platform/darwin/src/MGLBackgroundStyleLayer.mm

```diff
@@ -138,3 +139,20 @@ - (MGLTransition)backgroundPatternTransition {
 }
 
 @end
+
+@implementation MGLBackgroundStyleLayerFactory {
+    mbgl::style::BackgroundLayerFactory _rawLayerFactory;
+}
+
+- (MGLStyleLayer *)createWithRawLayer:(mbgl::style::Layer *)rawLayer
+{
+    NSParameterAssert(rawLayer);
+    return [[MGLBackgroundStyleLayer alloc] initWithRawLayer:rawLayer];
```

Comment:
In general, Objective-C design discourages the use of both singletons and factory classes. mbgl has multiple factory classes, but it may not be necessary to mirror that architecture one-for-one in Objective-C, since all these files are Objective-C++. (Note that abstract classes are found in Objective-C but are rather informal compared to in C++ or Java.)

At a glance, it seems like `LayerManagerMac` is acting as a bottleneck for layer creation of any type, allowing for a sort of central registry for arbitrary layer types. Consider folding the …StyleLayerFactory classes into their respective …StyleLayer classes and MGLStyleLayerManager into MGLStyleLayer. `-[MGLBackgroundStyleLayerFactory createWithRawLayer:]` would become `+[MGLBackgroundStyleLayer styleLayerWithRawLayer:]`. The registry in `_factories` can become a vector of `Class`es or even class names rather than an NSArray of factory class instances.